pypesto/objective/history.py

```diff
@@ -264,33 +283,213 @@ def update(
 
     def finalize(self):
-        """Finalize history. Called after a run."""
-        self.save_trace(finalize=True)
+        pass
 
-    def update_counts(self,
-                      sensi_orders: Tuple[int, ...],
-                      mode: str):
+    def _update_counts(self,
+                       sensi_orders: Tuple[int, ...],
+                       mode: str):
         """
         Update the counters.
         """
         if mode == MODE_FUN:
             if 0 in sensi_orders:
-                self.n_fval += 1
+                self._n_fval += 1
             if 1 in sensi_orders:
-                self.n_grad += 1
+                self._n_grad += 1
             if 2 in sensi_orders:
-                self.n_hess += 1
+                self._n_hess += 1
         elif mode == MODE_RES:
             if 0 in sensi_orders:
-                self.n_res += 1
+                self._n_res += 1
             if 1 in sensi_orders:
-                self.n_sres += 1
+                self._n_sres += 1
+
+    @property
+    def n_fval(self) -> int:
+        return self._n_fval
 
-    def init_trace(self, x: np.ndarray):
+    @property
+    def n_grad(self) -> int:
+        return self._n_grad
+
+    @property
+    def n_hess(self) -> int:
+        return self._n_hess
+
+    @property
+    def n_res(self) -> int:
+        return self._n_res
+
+    @property
+    def n_sres(self) -> int:
+        return self._n_sres
+
+    @property
+    def start_time(self) -> float:
+        return self._start_time
+
+
+class MemoryHistory(History):
+    """Tracks numbers of function evaluations and keeps an in-memory
+    trace of function evaluations.
+
+    Parameters
+    ----------
+    options:
+        History options.
+    """
+
+    def __init__(self, options: Union[HistoryOptions, Dict] = None):
+        super().__init__(options=options)
+        self._trace_keys = {X, FVAL, GRAD, HESS, RES, SRES, CHI2, SCHI2, TIME}
+        self._trace: Dict[str, Any] = {key: [] for key in self._trace_keys}
+
+    def update(
+            self,
+            x: np.ndarray,
+            sensi_orders: Tuple[int, ...],
+            mode: str,
+            result: ResultType
+    ) -> None:
+        super().update(x, sensi_orders, mode, result)
+        self._update_trace(x, sensi_orders, mode, result)
+
+    def _update_trace(self, x, sensi_orders, mode, result):
+        """Update internal trace representation."""
+        ret = extract_values(sensi_orders, mode, result, self.options)
+        for key in self._trace_keys - {X, TIME}:
+            self._trace[key].append(ret[key])
+        used_time = time.time() - self._start_time
+        self._trace[X].append(x)
+        self._trace[TIME].append(used_time)
+
+    def get_x_trace(self) -> Sequence[np.ndarray]:
+        return self._trace[X]
+
+    def get_fval_trace(self) -> Sequence[float]:
+        return self._trace[FVAL]
+
+    def get_grad_trace(self) -> Sequence[np.ndarray]:
```

Comment:
can integrate the Sequence part as additial dimension in `np.ndarray`, thats what you should currently get from the dataframe in CSVHistory.

Reply:
I think in general I would allow arbitrary return types, as parsing to numpy can take some time, and if all the requester wants is loop over the array, or select some value, then that's wasted effort. Do you know though whether one can specify dimensions or data types of numpy arrays? E.g. np.ndarray[float].

Reply:
https://github.com/ramonhagenaars/nptyping